packages/react-dom/src/__tests__/ReactDOMComponent-test.js

```diff
@@ -1059,17 +1065,37 @@ describe('ReactDOMComponent', () => {
         }),
       });
 
-      ReactDOM.render(<div checked={true} />, container);
+      ReactDOM.render(
+        <input type="checkbox" onChange={onChange} checked={true} />,
+        container,
+      );
       expect(nodeValueSetter).toHaveBeenCalledTimes(0);
 
-      ReactDOM.render(<div />, container);
+      expect(() => {
+        ReactDOM.render(
+          <input type="checkbox" onChange={onChange} />,
+          container,
+        );
+      }).toErrorDev(
+        'A component is changing a controlled input to be uncontrolled. This is likely caused by ' +
+          'the value changing from a defined to undefined, which should not happen. Decide between ' +
+          'using a controlled or uncontrolled input element for the lifetime of the component.',
+      );
       expect(nodeValueSetter).toHaveBeenCalledTimes(1);
 
-      ReactDOM.render(<div checked={false} />, container);
-      expect(nodeValueSetter).toHaveBeenCalledTimes(2);
-
-      ReactDOM.render(<div checked={true} />, container);
+      ReactDOM.render(
+        <input type="checkbox" onChange={onChange} checked={false} />,
+        container,
+      );
+      // TODO: Non-null values are updated twice on inputs. This is should ideally be fixed.
```

Comment:
ohh because we have generic logic and specific logic for input, got it.
I don't think we need the generic one, just the specific one: the generic one doesn't re-set the value if it shouldn't have changed, but we need to do that for the controlled components. that's why it sets it unconditionally in `updateChecked`